PR to merge all work from NCSU Senior Design project

dan144 opened the pull request:
This pull request contains all of the work from the NC State University ECE Senior Design team. The major features added include string search, batch processing, and OCR.

jazzido commented:
Wow, @dan144 ! This is amazing. Congrats on your senior project :) It seems that you all put a lot of work into this.

We are glad to work with you in merging this to tabula-java master, but we will need to go over some issues. First off, the [continuous integration job is failing](https://travis-ci.org/tabulapdf/tabula-java/jobs/245570025). When you get a chance, please update this branch with a fix for that issue and we can go from there.

Thanks again!

jazzido commented on src/main/java/technology/tabula/detectors/StringSearch.java:
```diff
@@ -0,0 +1,691 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.cli.ParseException;
```
No need for `org.apache.commons.cli.ParseException`. None of the methods in this class actually throw that exception, despite their signature.

jazzido commented on src/main/java/technology/tabula/detectors/StringSearch.java:
```diff
@@ -0,0 +1,691 @@
+import technology.tabula.Rectangle;
+import technology.tabula.TextElement;
+
+public class StringSearch {
```
What does this class do? How is it used?

jazzido commented on pom.xml:
```diff
@@ -279,6 +279,12 @@
         <version>2.8.0</version>
     </dependency>
+
+    <dependency>
```
Does `tess4j` include binaries for all the platforms that we support? (Windows, Mac and Linux). Does it require that Tesseract is present in the machine that Tabula is used?

criztovyl replied:
As far as I researched only Windows binaries are included.
[[1]](http://tess4j.sourceforge.net/tutorial/) [[2]](https://search.maven.org/remotecontent?filepath=net/sourceforge/tess4j/tess4j/3.4.0/tess4j-3.4.0.jar)

dan144 replied:
I will look into this. I believe @jeremybmerrill one time mentioned that Unix systems require a terminal install command for a `tesseract` package, but I'll investigate this and find the fix.

dan144 replied:
@dbangera23

dbangera23 replied:
Looks like we might need to add a line in the instructions for the linux/mac users. We used Tess4j in the project (http://tess4j.sourceforge.net/usage.html). which points to the original tesseract-ocr github page (https://github.com/tesseract-ocr/tesseract/wiki) that suggests running "sudo port install tesseract".

Although to be honest I'm not able to test since I don't have access to a linux device. Let me know if there is anything I can do and what you guys find out.

jeremybmerrill replied:
@dan144 @dbangera23 What does this do if tesseract isn't installed? This is probably something best solved with a detailed and intelligent error message.

dan144 replied:
[Per this answer](https://stackoverflow.com/a/30234912), I ran `sudo apt-get install tesseract-ocr` on a Ubuntu VM that had the failing test cases and it fixed them. It now passes all test on both my Windows machine and my Ubuntu VM (provided `tesseract-ocr` is installed). I don't know that this can be done as a built in dependency in Java.

It seems that the error is not a `TesseractException`, but rather an `UnsatisfiedLinkError`. I can add a catch for that and then log a message telling the user to run the command. Would that be an acceptable solution?

criztovyl replied:
I'm on Debian, will try to test :)

jazzido replied:
I'm trying to install the required deps on Travis (CI environment), getting this error:

```
testBatchExtractionWithOCR(technology.tabula.TestBatchExtractor)  Time elapsed: 4.324 sec  <<< ERROR!

java.lang.RuntimeException: Unable to load library 'gs': Native library (linux-x86-64/libgs.so) not found in resource path ([file:/home/travis/build/tabulapdf/tabula-java/target/surefire/surefirebooter8727177790678118125.jar])

Please download, install GPL Ghostscript from http://sourceforge.net/projects/ghostscript/files

and/or set the appropriate environment variable.

	at net.sourceforge.tess4j.util.PdfUtilities.convertPdf2Tiff(PdfUtilities.java:55)

	at technology.tabula.extractors.OcrConverter.extract(OcrConverter.java:39)

	at technology.tabula.extractors.BatchSelectionExtractor.extract(BatchSelectionExtractor.java:214)

	at technology.tabula.TestBatchExtractor.testBatchExtractionWithOCR(TestBatchExtractor.java:70)

	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

	at java.lang.reflect.Method.invoke(Method.java:498)

	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)

	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
...
```

What should we install on Ubuntu besides `tesseract-ocr`?

criztovyl replied:
Does this help? https://gist.github.com/criztovyl/358267e8375916ef6a8da404a3da5845

dan144 replied:
@jazzido it looks like we need ghostscript as well, which can be installed with `sudo apt-get install ghostscript`. That should be added to the `.travis.yml` file the same way you added the tesseract dependency I believe.

dan144 replied:
So the latest travis issue with Tesseract is `java.lang.UnsatisfiedLinkError: Error looking up function 'TessBaseAPICreate': /usr/lib/libtesseract.so.3.0.2: undefined symbol: TessBaseAPICreate`. When I run `apt-get install tesseract-ocr` on my local Ubuntu 16.04 system, it installs the same package, but with `libtesseract.so.3.0.4`, which contains the API calls for Tess4J that we used, which are present in Tesseract 3.0.3+. It looks like the Travis system is running Ubuntu 12.04LTS, which [ends support at 3.0.2](https://github.com/tesseract-ocr/tesseract/wiki/Compiling#leptonica). How do we want to proceed with this knowledge?

dan144 replied:
@jeremybmerrill this is the cause of the failures and my analysis of it. I don't know Travis super well, so I'm not sure how to fix it.

jeremybmerrill replied:
@dan144 @dbangera23 I see two failures when I run the tests myself (the other error may just be Travis being weird), both in testBatchExtractor, just with assertions that fail. Can you guys look into it and see what's going on, then get the tests to pass? Thanks in advance!

dbangera23 replied:
@jeremybmerrill I fixed the issue with testBatchExtractor. Looks like in the changes I forgot we had formatting in place for batch and now to keep in line with rest of Tabula, BatchWriter follows CSVWriter which messed up the expected output in testBatchExtractor.

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+
+	public int extract(String inputPath, String outputPath, String jsonPath, String processType, boolean ocrAllowed,
+			int overlapThreshold) {
+		System.out.println("OCR Selection: " + ocrAllowed);
```
Remove print to `stdout`. If needed, use a Logger. See example here: https://github.com/tabulapdf/tabula-java/blob/master/src/main/java/technology/tabula/ObjectExtractorStreamEngine.java#L96

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.cli.ParseException;
```
`org.apache.commons.cli.ParseException` is only used for indicating syntax errors when parsing command line arguments.

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+			File outputFile = new File(outputPath);
+
+			if (!inputFile.exists()) {
+				throw new ParseException("Input file or directory does not exist");
```
Use `java.io.FileNotFoundException`

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+			}
+
+			if (!outputFile.exists()) {
+				throw new ParseException("Output directory does not exist");
```
Use `java.io.FileNotFoundException`

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+			}
+
+			if (!outputFile.isDirectory()) {
+				throw new ParseException("Output path does not point to a directory");
```
Use `java.io.FileNotFoundException`

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+			}
+
+			if (parentPath.isDirectory()) {
+				System.out.println("Valid folder"); // check that path is valid
```
Do not print out to stdout.

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+					}
+
+					catch (Exception e) {
+						e.printStackTrace();
```
If swallowing all possible Exceptions, please explain why. This is likely incorrect. Also, do not print stack trace to console.

jazzido commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,473 @@
+			} // end of file list
+
+			if (!deleteList.isEmpty())
+				System.out.println("Deleting OCR'd files...");
```
Use a Logger.

jazzido commented on src/main/java/technology/tabula/extractors/OcrConverter.java:
```diff
@@ -0,0 +1,58 @@
+			// output - create PDF from image above
+			String image_filepath = image.getAbsolutePath();
+			if (OCR_rename) {
+				instance.createDocuments(image_filepath, input_filepath.substring(0, input_filepath.length() - 4) + "_OCR", list);
```
Do not pollute user folders. Write result to a temporary folder: http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createTempDirectory%28java.nio.file.Path,%20java.lang.String,%20java.nio.file.attribute.FileAttribute...%29

dan144 replied:
@dbangera23

jazzido commented on src/main/java/technology/tabula/extractors/OcrConverter.java:
```diff
@@ -0,0 +1,58 @@
+			}
+			System.out.println("OCR Done");
+			return "Success";
+		} catch (TesseractException e) {
```
Do not swallow exceptions, and don't print the stack trace to the console. If there was an exception, handle it properly (or throw it).

dan144 replied:
In the case of OcrConverter, an error simply means that the converter fails. Is it acceptable to log the error using Logger.error and return false? There's not much more handling that can be done for these errors (e.g. Security/IO, Tesseract errors are system problems outside the scope of Tabula).

jeremybmerrill replied:
If Tesseract fails, we should just pass it through to the user that it failed. (i.e. throw it in tabula-java, but then catch it in the tabula GUI to alert the user via the GUI)

dan144 replied:
@jeremybmerrill would it make sense to catch it, log here (like I currently do) and then re-throw it so it can be specially handled in the GUI to alert the user there? Also, now that I've seen it is also possible to get `UnsatisfiedLinkError`s for tesseract-related issues, should I handle that the same way we determine to handle this?

jazzido commented on src/main/java/technology/tabula/extractors/OcrConverter.java:
```diff
@@ -0,0 +1,58 @@
+import net.sourceforge.tess4j.util.LoadLibs;
+import net.sourceforge.tess4j.util.PdfUtilities;
+
+public class OcrConverter {
```
It would be great to expose OCR functionality in the [command-line app](https://github.com/tabulapdf/tabula-java/blob/master/src/main/java/technology/tabula/CommandLineApp.java)

dan144 replied:
@dbangera23

dbangera23 replied:
@jazzido Fixed. We used the command line argument of 'e'

jazzido commented on src/main/java/technology/tabula/extractors/OcrConverter.java:
```diff
@@ -0,0 +1,58 @@
+				instance.createDocuments(image_filepath, input_filepath.substring(0, input_filepath.length() - 4), list);
+			}
+			System.out.println("OCR Done");
+			return "Success";
```
Do not use a string as a return value.

dan144 commented:
I've addressed most of the comments here. Dean is looking into the remaining. Let us know what else we need to address, and thanks for all the feedback!

jeremybmerrill commented on src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java:
```diff
@@ -0,0 +1,430 @@
+// NOTES:
+//		need to remove tables from auto/spread list if they are used as a best guess
+//		or remove very similar tables from the list before extracting data
+public class BatchSelectionExtractor {
```
It might be worthwhile for this to mimic https://github.com/tabulapdf/tabula-java/blob/master/src/main/java/technology/tabula/extractors/SpreadsheetExtractionAlgorithm.java#L85 and have the `extract()` method return a list of tables, rather than writing the output to disk directly. 

I think it would also work for it to return a list of lists of tables (one list of tables per document in the batch job) or even a dictionary/hash mapping batch document filenames to a list of tables. 

This will also, I believe, solve one of the problems that's causing the continuous integration job to fail. Rather than requiring that an output directory exist, this `extract` method can "write" the output to a stream object, rather than a real file. That way, the tests aren't side-effecty in terms of creating files and folders.

Incidentally, I think there needs to be a little bit of work to deal with Linux/Mac compatibility. Where the test for this method is supposed to write some files to the `/src/test/resources/technology/tabula/batch/output/` directory, I end up with a file in `java/src/test/resources/technology/tabula/batch/` called `output\well_text_a.csv`. Java has the ability to let you name files without using `\` or `/` and letting Java figure out the right one to use.

dan144 replied:
I believe the last section of this is caused by [this section of the code.](https://github.com/dbangera23/tabula-java/blob/master/src/main/java/technology/tabula/extractors/BatchSelectionExtractor.java#L252) I'll switch this to use [Paths.get](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Paths.html#get(java.lang.String,%20java.lang.String...)).

dbangera23 replied:
Hey @jeremybmerrill, I changed the code to make the extract method close to how spreadsheetExtractionAlgorithm. It now returns a Map<String fileName, List<Table>> format. I also changed some of the code to make it modular. Now batch processing is done separately than writing. 

Let me know if there is anything else.
Thanks.

jeremybmerrill replied:
Oh, yes. The type signature looks good. Let's figure out why CI is failing, then, since @jazzido approved these too, we can see about getting these merged.

jeremybmerrill commented:
Okeedoke. Thanks for fixing the tests ("good failures" the ones where the tests fail because extraction got improved are common and kinda funny). I'm good to merge this. @jazzido ?

jazzido commented:
Hi @dan and @dbangera23,

Ran into another issue when playing with your branch. This command:

```
java -jar target/tabula-1.0.2-SNAPSHOT-jar-with-dependencies.jar -e -p 10 -g ~/Downloads/170803_crosstabs_Politico_v1_TB-1.pdf
Exception in thread "main" java.lang.UnsatisfiedLinkError: Unable to load library 'tesseract': Native library (darwin/libtesseract.dylib) not found in resource path ([file:/Users/manuel/Work/code/tabula-java/target/tabula-1.0.2-SNAPSHOT-jar-with-dependencies.jar])
	at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:271)
	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:398)
	at com.sun.jna.Library$Handler.<init>(Library.java:147)
	at com.sun.jna.Native.loadLibrary(Native.java:412)
	at com.sun.jna.Native.loadLibrary(Native.java:391)
	at net.sourceforge.tess4j.util.LoadLibs.getTessAPIInstance(LoadLibs.java:75)
	at net.sourceforge.tess4j.TessAPI.<clinit>(TessAPI.java:42)
	at net.sourceforge.tess4j.Tesseract.init(Tesseract.java:368)
	at net.sourceforge.tess4j.Tesseract.createDocuments(Tesseract.java:524)
	at net.sourceforge.tess4j.Tesseract.createDocuments(Tesseract.java:507)
	at technology.tabula.extractors.OcrConverter.extract(OcrConverter.java:45)
	at technology.tabula.CommandLineApp.extractFileTables(CommandLineApp.java:124)
	at technology.tabula.CommandLineApp.extractTables(CommandLineApp.java:104)
	at technology.tabula.CommandLineApp.main(CommandLineApp.java:74)
```

  - Even though I indicated that only page 10 needs to be processed, the app generated PNG images for *all* pages in the same folder that contains the input PDF. Of course, that should not happen. If you need to generated temporary files, they need to be generated in a temporary folder that is deleted afterwards.
  - As the trace shows, it failed. I'm on a Mac.

The document that I used for testing [is here](https://github.com/tabulapdf/tabula/files/1225798/tabula-170803_crosstabs_Politico_v1_TB-1.csv.zip), although any document will show this behavior.

Thanks!

jeremybmerrill commented:
Hey guys, does String Search have a way to use it from teh command line?

dbangera23 commented:
Hey @jazzido and @jeremybmerrill, Me and daniel will take a look at the mac error that you guys are seeing this weekend and see what we can do. I also didn't add in the functionality to do ocr on a per page basis. Just the whole document so we can look into that. 

Currently String search isn't implemented in the command line. We can look into making that work.

dan144 commented:
@jazzido I looked into your error on Mac, and I think it's the same issue we're seeing on Linux. Since the included JAR file only includes Windows-compatible Tess4J functions by default, I think you'll have manually install the Tesseract libraries. [Instructions are given here.](https://github.com/tesseract-ocr/tesseract/wiki#macos)

dbangera23 commented:
@jeremybmerrill Hey Jeremy, I looked into the option for adding a string based command line search and I don't think it's a good idea to do so at the moment. Maybe after the merge we can take a closer look. 

The problem is that CommandLineApp.whichArea(line); get the rectangle to determine where to process. After which extraction is done later at exactly those rectangles. The string search might return a rectangle that is different from one page to another. I could dynamically search for the rectangle in extractFile but even then String search can return multiple rectangles per page. 

I'm not sure how we want to handle this since the above "solution" would break how modular the code is and would need a redesign of the page class. Might be better to handle this functionality at a later date.

dbangera23 commented:
I'm finishing up the OCR per page functionality. Should have it done soon.

rosenjcb commented:
Was this pull request ever merged into the master branch? I've thought about adding OCR (tess4j) functionality but it looks like it was already attempted.

dan144 commented:
Hey @rosenjcb, this PR has not been worked on in some time. We had trouble with the Ubuntu version used for automated testing in this repo. No dev work has been done on this project in about 16 months, so if you're interested in picking up OCR effort on this project, this certainly shouldn't stop you.
